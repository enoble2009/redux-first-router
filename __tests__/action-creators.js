import { createMemoryHistory } from 'history'

import historyCreateAction from '../src/action-creators/historyCreateAction'
import middlewareCreateAction
  from '../src/action-creators/middlewareCreateAction'
import redirect from '../src/action-creators/redirect'
import { NOT_FOUND } from '../src/index'

it('historyCreateAction() - returns action created when history/address_bar chanages', () => {
  const history = createMemoryHistory()
  const pathname = '/info/foo'
  const prevLocation = { pathname: '/prev', type: 'PREV', payload: {} }
  const kind = 'pop'
  const routesMap = {
    INFO: '/info',
    INFO_PARAM: '/info/:param'
  }

  const action = historyCreateAction(
    pathname,
    routesMap,
    prevLocation,
    history,
    kind
  ) /*? $.meta.location */

  expect(action).toMatchSnapshot()

  expect(action.type).toEqual('INFO_PARAM')
  expect(action.payload).toEqual({ param: 'foo' })

  expect(action.type).toEqual(action.meta.location.current.type)
  expect(action.payload).toEqual(action.meta.location.current.payload)

  expect(action.meta.location.prev).toEqual(prevLocation)
  expect(action.meta.location.current).toEqual({
    pathname,
    type: 'INFO_PARAM',
    payload: { param: 'foo' }
  })

  expect(action.meta.location.kind).toEqual('pop')
})

it('middlewareCreateAction() - returns action created when middleware detects connected/matched action.type', () => {
  const history = createMemoryHistory()
  const receivedAction = { type: 'INFO_PARAM', payload: { param: 'foo' } }
  const routesMap = {
    INFO: '/info',
    INFO_PARAM: '/info/:param'
  }
  const prevLocation = { pathname: '/prev', type: 'PREV', payload: {} }

  const action = middlewareCreateAction(
    receivedAction,
    routesMap,
    prevLocation,
    history
  ) /*? $.meta.location */

  expect(action.type).toEqual('INFO_PARAM')
  expect(action.payload).toEqual({ param: 'foo' })

  expect(action.type).toEqual(action.meta.location.current.type)
  expect(action.payload).toEqual(action.meta.location.current.payload)

  expect(action.meta.location.prev).toEqual(prevLocation)
  expect(action.meta.location.current).toEqual({
    pathname: '/info/foo',
    type: 'INFO_PARAM',
    payload: { param: 'foo' }
  })

  expect(action.meta.location.kind).toEqual('push')

  expect(action).toMatchSnapshot()
})

it('middlewareCreateAction() - [action not matched to any routePath]', () => {
  const history = createMemoryHistory()
  const receivedAction = { type: 'BLA', payload: { someKey: 'foo' } }
  const routesMap = {
    INFO: '/info',
    INFO_PARAM: '/info/:param'
  }
  const prevLocation = { pathname: '/prev', type: 'PREV', payload: {} }

  const action = middlewareCreateAction(
    receivedAction,
    routesMap,
    prevLocation,
    history,
    '/not-found'
  ) /*? $.meta.location */

  expect(action.type).toEqual(NOT_FOUND)
  expect(action.payload).toEqual({ someKey: 'foo' })

  expect(action.meta.location.prev).toEqual(prevLocation)
  expect(action.meta.location.current.pathname).toEqual('/not-found')

  expect(action).toMatchSnapshot()
})

it('redirect(action) - sets action.meta.location.kind === "redirect"', () => {
  const receivedAction = { type: 'ANYTHING' }
  const action = redirect(receivedAction) /*? */

  expect(action.meta.location.kind).toEqual('redirect')
})
