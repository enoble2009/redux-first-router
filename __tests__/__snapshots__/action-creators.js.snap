// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`historyCreateAction() - returns action created when history/address_bar chanages 1`] = `
Object {
  "meta": Object {
    "location": Object {
      "current": Object {
        "pathname": "/info/foo",
        "payload": Object {
          "param": "foo",
        },
        "type": "INFO_PARAM",
      },
      "history": undefined,
      "kind": "pop",
      "prev": Object {
        "pathname": "/prev",
        "payload": Object {},
        "type": "PREV",
      },
    },
  },
  "payload": Object {
    "param": "foo",
  },
  "type": "INFO_PARAM",
}
`;

exports[`middlewareCreateAction() - [action not matched to any routePath] 1`] = `
Object {
  "meta": Object {
    "location": Object {
      "current": Object {
        "pathname": "/not-found",
        "payload": Object {
          "someKey": "foo",
        },
        "type": "@@redux-first-router/NOT_FOUND",
      },
      "history": undefined,
      "kind": undefined,
      "prev": Object {
        "pathname": "/prev",
        "payload": Object {},
        "type": "PREV",
      },
    },
  },
  "payload": Object {
    "someKey": "foo",
  },
  "type": "@@redux-first-router/NOT_FOUND",
}
`;

exports[`middlewareCreateAction() - returns action created when middleware detects connected/matched action.type 1`] = `
Object {
  "meta": Object {
    "location": Object {
      "current": Object {
        "pathname": "/info/foo",
        "payload": Object {
          "param": "foo",
        },
        "type": "INFO_PARAM",
      },
      "history": undefined,
      "kind": "push",
      "prev": Object {
        "pathname": "/prev",
        "payload": Object {},
        "type": "PREV",
      },
    },
  },
  "payload": Object {
    "param": "foo",
  },
  "type": "INFO_PARAM",
}
`;
